### PR TITLE
Revamp fixtures upload error detection

### PR DESCRIPTION
**Pull request by dannyroberts**
This will be much easier to review commit by commit because of initial file reorganization and splitting I did to be able to work with the code more easily.

This PR does the following:
- splits out validation from execution (there's still some inline errors that can be shown during execution, but I think all of the ones that you can validate by from excel file without looking at the database have been moved)
- validates synchronously and executes only if the file passes validation
- adds tests for fixture uploads, including a test per known error type with a minimal excel file exhibiting that error
- adds a test that tests that all the known errors are tested (meta, I know). If you add another check for a new known error, the build will fail until you test it (or work around it). I tried to set up the tests so that adding a new one is dead simple and doesn't require almost any boilerplate.
- In the fixture upload UI (and in fixture upload API), display _all_ errors encountered during validation, not just the first one. (There are still a few errors that stop further validation. I guess you could say it's because if the types sheet isn't correct itself, it's hard to validate the rest.)
- Fixes the bug originally causing http://manage.dimagi.com/default.asp?238734
- Fixes typos and various other minor defects I found along the way

### CODE REVIEW

File: corehq/apps/fixtures/upload/__init__.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from .upload import (
```
Comment (dimagimon) [inline corehq/apps/fixtures/upload/__init__.py]: F401 'DELETE_HEADER' imported but unused
F401 'FixtureWorkbook' imported but unused
F401 'get_workbook' imported but unused
F401 'run_upload' imported but unused
F401 'safe_fixture_upload' imported but unused
F401 'validate_file_format' imported but unused

**Comment by dimagimon** [comment]
There are 10 errors:

* corehq/apps/fixtures/upload/run_upload.py, line 63 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 67 - F841 local variable 'e' is assigned to but never used
* corehq/apps/fixtures/upload/run_upload.py, line 120 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 144 - E501 line too long (138 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 150 - E501 line too long (126 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 156 - E501 line too long (139 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 233 - E501 line too long (121 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 20 - E501 line too long (129 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 101 - W601 .has_key() is deprecated, use 'in'
* corehq/apps/fixtures/upload/upload.py, line 166 - E501 line too long (116 > 115 characters)

**Comment by dimagimon** [comment]
There are 10 errors:

* corehq/apps/fixtures/upload/run_upload.py, line 63 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 67 - F841 local variable 'e' is assigned to but never used
* corehq/apps/fixtures/upload/run_upload.py, line 120 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 144 - E501 line too long (138 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 150 - E501 line too long (126 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 156 - E501 line too long (139 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 233 - E501 line too long (121 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 20 - E501 line too long (129 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 101 - W601 .has_key() is deprecated, use 'in'
* corehq/apps/fixtures/upload/upload.py, line 166 - E501 line too long (116 > 115 characters)

**Comment by dimagimon** [comment]
There are 15 errors:

* corehq/apps/fixtures/upload/run_upload.py, line 63 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 67 - F841 local variable 'e' is assigned to but never used
* corehq/apps/fixtures/upload/run_upload.py, line 120 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 144 - E501 line too long (138 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 150 - E501 line too long (126 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 156 - E501 line too long (139 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 195 - E501 line too long (120 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 197 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 203 - E501 line too long (120 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 205 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 228 - E501 line too long (128 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 251 - E501 line too long (130 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 20 - E501 line too long (129 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 101 - W601 .has_key() is deprecated, use 'in'
* corehq/apps/fixtures/upload/upload.py, line 166 - E501 line too long (116 > 115 characters)

**Comment by dimagimon** [comment]
There are 17 errors:

* corehq/apps/fixtures/tests/test_upload.py, line 17 - E501 line too long (122 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 63 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 67 - F841 local variable 'e' is assigned to but never used
* corehq/apps/fixtures/upload/run_upload.py, line 120 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 144 - E501 line too long (138 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 150 - E501 line too long (126 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 156 - E501 line too long (139 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 201 - E501 line too long (120 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 203 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 209 - E501 line too long (120 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 211 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 234 - E501 line too long (128 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 257 - E501 line too long (130 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 20 - E501 line too long (129 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 59 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 104 - W601 .has_key() is deprecated, use 'in'
* corehq/apps/fixtures/upload/upload.py, line 169 - E501 line too long (116 > 115 characters)

**Comment by dimagimon** [comment]
There are 17 errors:

* corehq/apps/fixtures/tests/test_upload.py, line 17 - E501 line too long (122 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 64 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 68 - F841 local variable 'e' is assigned to but never used
* corehq/apps/fixtures/upload/run_upload.py, line 121 - E501 line too long (134 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 145 - E501 line too long (138 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 151 - E501 line too long (126 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 157 - E501 line too long (139 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 202 - E501 line too long (120 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 204 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 210 - E501 line too long (120 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 212 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 235 - E501 line too long (128 > 115 characters)
* corehq/apps/fixtures/upload/run_upload.py, line 258 - E501 line too long (130 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 20 - E501 line too long (129 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 59 - E501 line too long (117 > 115 characters)
* corehq/apps/fixtures/upload/upload.py, line 104 - W601 .has_key() is deprecated, use 'in'
* corehq/apps/fixtures/upload/upload.py, line 169 - E501 line too long (116 > 115 characters)

**Comment by dannyroberts** [comment]
@czue / anyone I think this is good to review now. It's big enough that I am going to throw it up on staging for a bit and do a little testing with some more real-life files, but I'm pretty happy with the level of testing and think I'm unlikely to find anything blocking

File: corehq/apps/fixtures/upload/run_upload.py [hunk 2]
```diff
@@ -233,6 +166,90 @@ def _update_progress(table_count, item_count, items_in_table):
...
 
+def _diff_lists(list_a, list_b):
+    set_a = set(list_a)
+    set_b = set(list_b)
+    not_in_b = set_a.difference(set_b)
+    not_in_a = set_a.difference(set_a)
+    return list(not_in_a), list(not_in_b)
+
+
+def validate_fixture_upload(workbook):
```
Comment (czue) [inline corehq/apps/fixtures/upload/run_upload.py]: should this be reviewed or is it basically just moved from above?

File: corehq/apps/fixtures/upload/upload.py [hunk 3]
```diff
@@ -182,8 +182,9 @@ def get_all_type_sheets(self):
         return type_sheets
 
     def validate(self):
+        from corehq.apps.fixtures.upload.run_upload import validate_fixture_upload
         self.get_types_sheet()
-        self.get_all_type_sheets()
```
Comment (czue) [inline corehq/apps/fixtures/upload/upload.py]: was this just running a check that was also run in what became `validate_fixture_upload`?

File: corehq/apps/fixtures/upload/run_upload.py [hunk 4]
```diff
@@ -170,7 +170,7 @@ def _diff_lists(list_a, list_b):
     set_a = set(list_a)
     set_b = set(list_b)
     not_in_b = set_a.difference(set_b)
-    not_in_a = set_a.difference(set_a)
+    not_in_a = set_b.difference(set_a)
```
Comment (czue) [inline corehq/apps/fixtures/upload/run_upload.py]: sneaky

File: corehq/apps/fixtures/upload/run_upload.py [hunk 5]
```diff
@@ -183,8 +183,11 @@ def validate_fixture_upload(workbook):
         tag = table_def.table_id
         fields = table_def.fields
         item_attributes = table_def.item_attributes
-        data_items = list(workbook.get_data_sheet(tag))
-        for sort_key, data_item in enumerate(data_items):
+        try:
+            data_item = iter(workbook.get_data_sheet(tag)).next()
```
Comment (czue) [inline corehq/apps/fixtures/upload/run_upload.py]: can you explain why this no longer needs to loop through all of the `data_items`?

File: corehq/apps/fixtures/tests/test_upload.py [hunk 6]
```diff
@@ -1,24 +1,36 @@
...
 
-dot = os.path.dirname(__file__)
 
-UPLOAD_FILE = os.path.join(dot, 'test_upload', 'incorrect_fixtures.xlsx')
-ERRORS = [
+def make_path(args):
+    return os.path.join(os.path.dirname(__file__), *args)
+
+
+class UploadTestConfig(namedtuple('UploadTestConfig', ['upload_file_', 'error_messages'])):
```
Comment (czue) [inline corehq/apps/fixtures/tests/test_upload.py]: is the trailing `_` a convention you're proposing/copying?

Comment (czue) [inline corehq/apps/fixtures/tests/test_upload.py]: (nvm see this changed later)

File: corehq/apps/fixtures/upload/upload.py [hunk 7]
```diff
@@ -41,15 +41,15 @@
         "'{property}' of the field '{field}' as specified in its 'types' definition"
     ),
     "sheet_has_extra_property": ugettext_noop(
-        "Excel-sheet '{tag}'' has an extra property "
+        "Excel-sheet '{tag}' has an extra property "
```
Comment (czue) [inline corehq/apps/fixtures/upload/upload.py]: :bird: :eye:

File: corehq/apps/fixtures/tests/test_upload.py [hunk 8]
```diff
@@ -29,10 +30,32 @@ class TestFixtureUpload(SimpleTestCase):
...
+                wb.validate()
+            self.assertEqual(context.exception.errors, config.error_messages)
+        else:
+            # assert doesn't raise anything
             wb.validate()
-        self.assertEqual(context.exception.errors, config.error_messages)
 
+    test_ok = _upload_test('ok', [])
+
+    def test_comprehensiveness(self):
```
Comment (czue) [inline corehq/apps/fixtures/tests/test_upload.py]: if you wanted to get even more meta you could add a test for this test. :recycle:

File: corehq/apps/fixtures/upload/upload.py [hunk 9]
```diff
@@ -54,10 +61,6 @@
     "wrong_field_property_combos": ugettext_noop(
         "Number of values for field '{field}' and attribute '{prop}' should be same"
     ),
-    "replace_with_UID": ugettext_noop(
```
Comment (czue) [inline corehq/apps/fixtures/upload/upload.py]: was this just never referenced?

**Review by czue — APPROVED** [review]
looks great. just a few random questions (i'm guessing everything has an obvious answer and just wasn't obvious to me)

File: corehq/apps/fixtures/upload/run_upload.py (hunk 5 above)
Comment (dannyroberts) [inline corehq/apps/fixtures/upload/run_upload.py]: Yeah, this isn't super obvious, but the only thing being validated is the header format for each of the sheets; however, we don't have direct access to the header row. Instead, the header row has been used to format each data row into this (deserialized) json `data_item`. That means that at this point the only way to get at the header format is through one of the rows—any of the rows. That's why we only need to look at the first one.

File: corehq/apps/fixtures/upload/run_upload.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/fixtures/upload/run_upload.py]: It was basically moved in the first commit touching this, but then it was pretty heavily edited, so I'd review this part commit by commit

File: corehq/apps/fixtures/upload/run_upload.py (hunk 5 above)
Comment (czue) [inline corehq/apps/fixtures/upload/run_upload.py]: gotcha. could be worth a comment explaining that.

File: corehq/apps/fixtures/upload/run_upload.py (hunk 2 above)
Comment (czue) [inline corehq/apps/fixtures/upload/run_upload.py]: cool, yeah this was a commit comment, but i guess that doesn't show up easily

File: corehq/apps/fixtures/tests/test_upload.py (hunk 6 above)
Comment (dannyroberts) [inline corehq/apps/fixtures/tests/test_upload.py]: `namedtuple` doesn't let you prefix a value with `_` so I did this for speed while I was iterating on it, though I don't love it as a convention and eventually changed it.

File: corehq/apps/fixtures/tests/test_upload.py (hunk 8 above)
Comment (dannyroberts) [inline corehq/apps/fixtures/tests/test_upload.py]: "`self.assert_(hasattr(self, 'test_comprehensiveness'))`"

Comment (czue) [inline corehq/apps/fixtures/tests/test_upload.py]: haha, i meant add an error type and make sure it fails but i was kidding

**Comment by dannyroberts** [comment]
I did a couple uploads and the ones that should fail with good (ahem, at least better) error messages and the ones that should work work. Good to merge on my end!

**Comment by czue** [comment]
looks good on :white_check_mark:

**Comment by dannyroberts** [comment]
✅

**Pull request merged**